File: src/pset2/vigenere/vigenere.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "../../../lib/cs50.h"

int isAlpha(char const *str);
int shift(char c);

int main(int argc, char const *argv[])
{
    // Check argument count
    if (argc < 2 || argc > 2 || !isAlpha(argv[1]))
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }

    // Plain text
    char *p = get_string("plaintext: ");

    // Secret key
    char const *ks = argv[1];
    int kslen = strlen(ks);
    int k[kslen];
    for (int i = 0; i < kslen; i++)
    {
        k[i] = shift(ks[i]);
    }

    // Secret key counter
    int ki = 0;

    // Cipher text
    char *c = p;

    // Chesar:
    for (int i = 0, n = strlen(p); i < n; i++)
    {
        if (ki >= kslen)
        {
            ki = 0;
        }

        if (isalpha(p[i]))
        {
            if (isupper(c[i]))
            {
                c[i] = ((p[i] + k[ki] - 65) % 26) + 65;
            }
            else
            {
                c[i] = ((p[i] + k[ki] - 97) % 26) + 97;
            }
            ki++;
        }
        else
        {
            c[i] = p[i];
        }
    }

    // Print cipher text
    printf("ciphertext: %s\n", c);

    // Return success result
    return 0;
}

int isAlpha(char const *str)
{
    for (int i = 0; i < strlen(str); i++)
    {
        if (!isalpha(str[i]))
        {
            return 0;
        }
    }

    return 1;
}

int shift(char c)
{
    if (isupper(c))
        return c - 65;
    else if (islower(c))
        return c - 97;
    return c;
}

File: src/pset6/vigenere/vigenere.py
from sys import argv
from cs50 import get_string

# Get argument count
argc = int(len(argv))

# Initialization point of program
def main() -> int:
    # Check arguments
    if argc != 2 or argv[1].isalpha() == False:
        print("Usage: python3 vigenere.py key")
        return 1
    
    # Get key
    key = argv[1]
    key_len = len(key)

    # Shift all key chars
    key = shiftall(key)

    # Get plaintext
    plaintext = get_string("plaintext: ")

    # Define ciphertext string
    ciphertext = ""

    # Define key counter
    key_counter = 0
    
    # Encrypt plaintext
    for i in range(len(plaintext)):
        if key_counter >= key_len:
            key_counter = 0
            pass
        if plaintext[i].isalpha():
            if plaintext[i].isupper():
                ciphertext = ciphertext + chr(((ord(plaintext[i]) + ord(key[key_counter]) - 65) % 26) + 65)
            else:
                ciphertext = ciphertext + chr(((ord(plaintext[i]) + ord(key[key_counter]) - 97) % 26) + 97)
            key_counter = key_counter + 1
        else:
            ciphertext = ciphertext + plaintext[i]

    # Print the ciphertext
    print(f"ciphertext: {ciphertext}")   

    # Return success result
    return 0

def shiftall(s) -> str:
    r = ""
    for i in range(len(s)):
        if s[i].isupper():
            r = r + chr(ord(s[i]) - 65)
        elif s[i].islower():
            r = r + chr(ord(s[i]) - 97)
    return r

# Call main function
if __name__ == "__main__":
    exit(main())